tokio/src/task/join_map.rs

```diff
@@ -0,0 +1,540 @@
+    /// scheduled to receive a wakeup.
+    ///
+    /// # Returns
+    ///
+    /// This function returns:
+    ///
+    ///  * `Poll::Pending` if the `JoinMap` is not empty but there is no task whose output is
+    ///     available right now.
+    ///  * `Poll::Ready(Some((key, Ok(value))))` if one of the tasks in this
+    ///    `JoinMap` has completed. The `value` is the return value of that
+    ///    task, and `key` is the key associated with the task.
+    ///  * `Poll::Ready(Some((key, Err(err)))` if one of the tasks in this
+    ///    `JoinMap` has panicked or been aborted. `key` is the key associated
+    ///    with the task that panicked or was aborted.
+    ///  * `Poll::Ready(None)` if the `JoinMap` is empty.
+    ///
+    /// Note that this method may return `Poll::Pending` even if one of the tasks has completed.
+    /// This can happen if the [coop budget] is reached.
+    ///
+    /// [coop budget]: crate::task#cooperative-scheduling
+    fn poll_join_one(&mut self, cx: &mut Context<'_>) -> Poll<Option<(K, Result<V, JoinError>)>> {
+        // The call to `pop_notified` moves the entry to the `idle` list. It is moved back to
+        // the `notified` list if the waker is notified in the `poll` call below.
+        let mut entry = match self.joins.pop_notified(cx.waker()) {
+            Some(entry) => entry,
+            None => {
+                if self.is_empty() {
+                    return Poll::Ready(None);
+                } else {
+                    // The waker was set by `pop_notified`.
+                    return Poll::Pending;
+                }
+            }
+        };
+
+        let res = entry.with_value_and_context(|(_, jh), ctx| Pin::new(jh).poll(ctx));
+
+        if let Poll::Ready(res) = res {
+            let (key, _) = entry.remove();
+            self.aborts.remove(&key);
+            Poll::Ready(Some((key, res)))
+        } else {
+            // A JoinHandle generally won't emit a wakeup without being ready unless
+            // the coop limit has been reached. We yield to the executor in this
+            // case.
+            cx.waker().wake_by_ref();
+            Poll::Pending
+        }
+    }
+}
+
+impl<K, V, S> JoinMap<K, V, S>
+where
+    K: 'static,
+    V: 'static,
+{
+    /// Aborts all tasks on this `JoinMap`.
+    ///
+    /// This does not remove the tasks from the `JoinMap`. To wait for the tasks to complete
+    /// cancellation, you should call `join_one` in a loop until the `JoinMap` is empty.
+    pub fn abort_all(&mut self) {
+        self.joins.for_each(|(_, jh)| jh.abort());
+    }
+
+    /// Removes all tasks from this `JoinMap` without aborting them.
+    ///
+    /// The tasks removed by this call will continue to run in the background even if the `JoinMap`
+    /// is dropped. They may still be aborted by key.
+    pub fn detach_all(&mut self) {
+        self.joins.drain(drop);
+    }
```

Comment:
This should also clear the map.

Reply:
I wasn't sure --- this way, the user can still abort those tasks by key, even if they have been detached. Is this something we want, or not?